be/src/exec/vectorized/json_scanner.cpp

```diff
@@ -345,55 +345,68 @@ Status JsonReader::close() {
+                // Expand outer array automatically according to _is_ndjson.
+                if (_is_ndjson) {
+                    RETURN_IF_ERROR(_read_chunk<ExpandedJsonDocumentStreamParserWithRoot>(
+                            chunk, &rows_to_read, &reordered_slot_descs, &is_reordered));
+                } else {
+                    RETURN_IF_ERROR(_read_chunk<ExpandedJsonArrayParserWithRoot>(chunk, &rows_to_read,
+                                                                                 &reordered_slot_descs, &is_reordered));
+                }
             } else {
-                return _read_chunk<ExpandedJsonArrayParserWithRoot>(chunk, rows_to_read, slot_descs);
+                if (_is_ndjson) {
+                    RETURN_IF_ERROR(_read_chunk<JsonDocumentStreamParserWithRoot>(
+                            chunk, &rows_to_read, &reordered_slot_descs, &is_reordered));
+                } else {
+                    RETURN_IF_ERROR(_read_chunk<JsonArrayParserWithRoot>(chunk, &rows_to_read, &reordered_slot_descs,
+                                                                         &is_reordered));
+                }
             }
         } else {
-            if (_is_ndjson) {
-                return _read_chunk<JsonDocumentStreamParserWithRoot>(chunk, rows_to_read, slot_descs);
+            // Without json root set, the strip_outer_array determines whether to expand outer array.
+            if (_scanner->_strip_outer_array) {
+                RETURN_IF_ERROR(
+                        _read_chunk<JsonArrayParser>(chunk, &rows_to_read, &reordered_slot_descs, &is_reordered));
             } else {
-                return _read_chunk<JsonArrayParserWithRoot>(chunk, rows_to_read, slot_descs);
+                RETURN_IF_ERROR(_read_chunk<JsonDocumentStreamParser>(chunk, &rows_to_read, &reordered_slot_descs,
+                                                                      &is_reordered));
             }
         }
-    } else {
-        // Without json root set, the strip_outer_array determines whether to expand outer array.
-        if (_scanner->_strip_outer_array) {
-            return _read_chunk<JsonArrayParser>(chunk, rows_to_read, slot_descs);
-        } else {
-            return _read_chunk<JsonDocumentStreamParser>(chunk, rows_to_read, slot_descs);
-        }
-    }
+    } while (rows_to_read > 0);
+    return Status::OK();
 }
 
 template <typename ParserType>
-Status JsonReader::_read_chunk(Chunk* chunk, int32_t rows_to_read, const std::vector<SlotDescriptor*>& slot_descs) {
+Status JsonReader::_read_chunk(Chunk* chunk, int32_t* rows_to_read, std::vector<SlotDescriptor*>* slot_descs,
+                               bool* is_reordered) {
     simdjson::ondemand::object row;
 
     auto parser = down_cast<ParserType*>(_parser.get());
 
-    std::vector<SlotDescriptor*> reordered_slot_descs(slot_descs);
-    for (int32_t n = 0; n < rows_to_read; n++) {
+    size_t n = 0;
+    while ((*rows_to_read)-- > 0) {
```

Comment:
actually you can't change the rows_to_read because you don't read anything.
Why not return StatusOf<int32_t> when successfully read,  return the number of rows read